crates/binjs_io/src/escaped_wtf8.rs

```diff
@@ -277,5 +283,176 @@ pub fn for_print(s: &SharedString) -> SharedString {
         };
     }
 
-    SharedString::from_string(String::from_utf8(buf).expect("Escaped string should be valid UTF-8"))
+    String::from_utf8(buf).expect("Unescaped string should be valid UTF-8")
+}
+
+/// If the given `bytes` contains any unicode escape \uXXXX and that is lone
+/// surrogate, convert it to escaped lone surropgate.
+///
+/// This assumes the input is well-formed JSON, and panics otherwise.
+pub fn from_unicode_escape(s: String) -> String {
+    let pos = s.find(|c| c == char::from(b'\\'));
+
+    //   ...... \uXXXX ......
+    //   ^      ^
+    //   |      |
+    //   s     end
+    let mut end = if let Some(end) = pos {
+        end
+    } else {
+        return s;
+    };
+
+    let mut input = s.as_str().as_bytes();
+
+    let mut buf: Vec<u8> = Vec::with_capacity(input.len().next_power_of_two());
+    loop {
+        let head = &input[..end];
+        buf.extend_from_slice(head);
+
+        let c = input[end + 1];
+        if c == b'u' {
+            let c1 = input[end + 2];
```

Comment:
Looks like this block is repeated often.

Maybe you should factor out
```rust
let c1 = input[end + 2];
let c2 = input[end + 3];
let c3 = input[end + 4];
let c4 = input[end + 5];
let codepoint: u16 = (decode_hex_char(c1) << 12)
                | (decode_hex_char(c2) << 8)
                | (decode_hex_char(c3) << 4)
                | decode_hex_char(c4);
```

as a function?